### PR TITLE
Make AND/OR variadic.

### PR DESCRIPTION
**Pull request by ggevay**
Make AND/OR variadic, and translate `IN (1,2,3,4,5)` to a variadic `OR`. This avoids stack overflows that were arising due to the large OR chains. It also simplifies some code which was decomposing or creating chains of ANDs or ORs. It also makes the evaluation of AND/ORs with many operands slightly faster.

### Motivation

Which of the following best describes the motivation behind this PR?

  * This PR fixes a recognized bug:
    * #6246,
    * #6247

### Tips for reviewer

An IN list used to be desugared in `transform_ast.rs` into a chain of ORs at the AST level, which was translated into a chain of ORs in HIR. Now I removed the desugaring, and instead handle IN lists when translating from AST to HIR, creating a variadic OR directly (`plan_in_list`). (An alternative design would have been to create a variadic OR at the AST level and desugar IN lists to that, but I didn't do this because I guess it's better if the AST just keeps reflecting the actual SQL syntax.)

I made some refactoring in `reduce` in `expr/src/scalar/mod.rs`: I moved the things that were after the main loop into the main loop. Now it's ok if those things don't perform certain cleanups after themselves, but instead rely on other transforms in the main loop to clean up. For example, there is no need for an extra canonicalizing sort of AND/OR arguments inside or after `undistribute_and_or`, because the main loop will do that. Also, `undistribute_and_or` can create opportunities for simplifications by `reduce_and_canonicalize_and_or`, see doc comment. However, there is a slight risk of infinite loops appearing. To mitigate this risk, I will ask Philip to run his randomized tests when I move the PR out of draft (as we discussed in one of the hangout meetings). @philip-stoev, could you please test this? 

I added the handling of `CallVariadic` at some places which were only handling `CallUnary` and `CallBinary`, but not `CallVariadic`. (`propagates_null_from_subexpression`, `decompose_is_null`, `PredicatePushdown::can_inline`)

In `chbench.slt:1424` we lost a CSE. This is because `(a || b)` is no longer a subexpression of `((a || b) || c)`, because it gets flattened into `(a || b || c)`. We might recover this optimization when working on https://github.com/MaterializeInc/materialize/issues/7825.

In `extract_equal_or_both_null`, the canonical order of OR operands changed, so I added handling of both orders of OR operands. Alternatively, it could just rely on the new canonical order (as it did before on the old order), but that is a bit brittle: if the canonical order changes again due to some reason, then it wouldn't match, see https://materializeinc.slack.com/archives/C01BE3RN82F/p1658156355698269?thread_ts=1658148794.295369&cid=C01BE3RN82F

In the test in `reduce:768` I added an extra argument to `coalesce`. Earlier, it had the same complexity as the other expression, in which case whether it's used for a reduction depended on the sort order between the expressions. Now this sort order changed (because of the AND moving to `CallVariadic`), so the optimization was not happening in the test.

To avoid changing `a && b` in many tests to `&&(a, b)` I added `func.is_infix_op` handling in printing of `CallVariadic`. (`explain_new/hir/text.rs`, `sql/src/plan/explain.rs`, `query_model/model/scalar.rs`) @aalexandrov 

In `src/expr/src/linear.rs` I changed two prints from `{:?}` to `{}`, which you can see in action in `test/testdrive/temporal.td`.

I will open some issues later:
- Generalizing `flatten_and_or` to other associative variadic functions. (`concat`, `least`, `greatest`)
- Determinism and `propagates_nulls`, see https://materializeinc.slack.com/archives/C01BE3RN82F/p1657644478517709 and `scalar/mod.rs:1031`.
- Adding more cases to `VariadicFunc::introduces_nulls` and `BinaryFunc::introduces_nulls`.
- The `and true` weirdness mentioned in `src/sql/src/plan/query.rs:2891`.

### Checklist

- [x] This PR has adequate test coverage / QA involvement has been duly considered.
  - Large IN list: `in_list.slt`
  - `flatten_and_or`: `testdata/reduce:284`
  - `undistribute_and/or` with more than 2 args at the top level: `testdata/reduce:295`
  - `extract_equal_or_both_null` change (making it work for both operand orders of the OR): `testdata/predicate-pushdown` lines 159-160 swapped
- [x] This PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way) and therefore is tagged with a `T-protobuf` label.
- [x] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):

  -   No user-facing behavior changes.

### CODE REVIEW

File: src/sql/src/plan/lowering.rs [hunk 1]
```diff
@@ -792,17 +792,22 @@ impl HirScalarExpr {
...
+                            let mut else_inner = get_inner.filter(vec![SS::CallVariadic {
+                                func: mz_expr::VariadicFunc::Or,
+                                exprs: vec![
+                                    SS::CallBinary {
+                                        func: mz_expr::BinaryFunc::Eq,
+                                        expr1: Box::new(cond_expr.clone()),
+                                        expr2: Box::new(SS::literal_ok(
+                                            Datum::False,
+                                            ScalarType::Bool,
+                                        )),
```
Comment (wangandi) [inline src/sql/src/plan/lowering.rs:804]: opportunity to use the existing `MirScalarExpr::call_binary` and the new `MirScalarExpr::call_is_null` function here.

File: src/sql/src/plan/query.rs [hunk 2]
```diff
@@ -3066,10 +3081,12 @@ fn plan_and(
...
-    Ok(HirScalarExpr::CallBinary {
-        func: BinaryFunc::And,
-        expr1: Box::new(plan_expr(&ecx, left)?.type_as(&ecx, &ScalarType::Bool)?),
-        expr2: Box::new(plan_expr(&ecx, right)?.type_as(&ecx, &ScalarType::Bool)?),
+    Ok(HirScalarExpr::CallVariadic {
+        func: VariadicFunc::And,
+        exprs: vec![
+            plan_expr(&ecx, left)?.type_as(&ecx, &ScalarType::Bool)?,
+            plan_expr(&ecx, right)?.type_as(&ecx, &ScalarType::Bool)?,
+        ],
```
Comment (wangandi) [inline src/sql/src/plan/query.rs:3089]: opportunity to use the new `HirScalarExpr::variadic_and` function

**Review by wangandi — COMMENTED** [review]
Got through 20 out of the 32 files so far, but mostly the short ones.

**Comment by philip-stoev** [comment]
I am on it, please give me until Monday

**Comment by philip-stoev** [comment]
Item No 1. Cosmetic -- an In-list with 1 element on the right side looks like this in this branch:

```
materialize=> explain raw plan for select * from t1 where f1 in (1);
             Raw Plan             
----------------------------------
 %0 =                            +
 | Get materialize.public.t1 (u1)+
 | Filter ||((#0 = 1))           +
```

I think this should have been converted to a simple equality, or , at least, printed without the `||`

In main, it looks like this (even weirder):

```
materialize=>  explain raw plan for select * from t1 where f1 in (1);
             Raw Plan             
----------------------------------
 %0 =                            +
 | Get materialize.public.t1 (u1)+
 | Filter (false || (#0 = 1))    +
```

**Comment by ggevay** [comment]
> Item No 1. Cosmetic -- an In-list with 1 element on the right side looks like this in this branch

Fixed it by calling `HirScalarExpr::variadic_or`, which automatically simplifies if there is only 1 arg.

File: src/sql/src/plan/query.rs (hunk 2 above)
Comment (ggevay) [reply src/sql/src/plan/query.rs]: Done

File: src/sql/src/plan/query.rs [hunk 3]
```diff
@@ -2888,15 +2888,25 @@ fn plan_using_constraint(
...
-    let on = join_exprs
-        .into_iter()
-        .fold(HirScalarExpr::literal_true(), |expr1, expr2| {
-            HirScalarExpr::CallBinary {
-                func: BinaryFunc::And,
-                expr1: Box::new(expr1),
-                expr2: Box::new(expr2),
-            }
-        });
+    // TODO: This should be just
```
[the remark this answers is not in the document]
Comment (ggevay) [reply src/sql/src/plan/query.rs:2891]: Yes, it's on my todo list to file an issue, will do it a bit later.

File: src/sql/src/plan/lowering.rs (hunk 1 above)
Comment (ggevay) [reply src/sql/src/plan/lowering.rs]: Done

**Review by philip-stoev — APPROVED** [review]
In addition to manual experimentation, I created a randomised workload that was generating large IN expressions with various items on both sides (not just a column reference and literals, respectively) and then putting those IN expressions in various contexts, in particular IN being part of a larger conjunction/disjunction expression.

No wrong results or panics were detected. Some query plans for degenerate queries were slightly better, but nothing worth noting.

**Comment by philip-stoev** [comment]
Item No 2. I noticed that the order in which some predicates are printed is now different. I personally do not mind any particular order, as long as it is stable. 

Complete speculation: I noticed that you used `HashSet` in your code -- previously, we had to replace some uses of hash-based structures with btree-based to ensure that plans do not flip around.

There was also some work a long time ago to order predicates in a systematic way, e.g. https://github.com/MaterializeInc/materialize/pull/9795 . Not sure how important it is and if those rules are followed across the board.

**Comment by philip-stoev** [comment]
Here is an example: ```SELECT f3 FROM t1 WHERE  f3  IN ( 1,  f3 % f3  , f3) ```

The order was previously the syntacic one:

```
Filter (((#2 = 1) || (#2 = (#2 % #2))) || (#2 = #2))
```

With this PR, the order appears to be literals > column references > expressions:

```
Filter ((#2 = #2) || (#2 = 1) || (#2 = (#2 % #2)))
```

I think both are equally valid, as long as the outcome is not determined by the implementation of the hashmap.

**Comment by ggevay** [comment]
> Item No 2. I noticed that the order in which some predicates are printed is now different. I personally do not mind any particular order, as long as it is stable.

Should be mostly stable (at the stage where we print explains or execute queries). The code that takes care of this canonicalization is `src/expr/src/scalar/mod.rs`: `reduce_and_canonicalize_and_or`: `exprs.sort();`, which will sort the expressions based on their internal representations. Note that this sort order can sometimes change with big refactorings, e.g., now it changed because AND/OR was earlier called using `CallBinary` but now it's called using `CallVariadic`. But it won't change due to HashSet implementation changes, or due to changes in other parts of a query plan, or due to new transforms in scalar `reduce`, etc.

File: src/transform/src/predicate_pushdown.rs [hunk 4]
```diff
@@ -656,24 +658,28 @@ impl PredicatePushdown {
...
+                                                            func: UnaryFunc::IsNull(func::IsNull),
+                                                            expr: Box::new(expr2.clone()),
+                                                        },
+                                                        MirScalarExpr::CallUnary {
+                                                            func: UnaryFunc::IsNull(func::IsNull),
+                                                            expr: Box::new(expr1.clone()),
+                                                        },
+                                                    ],
+                                                },
+                                            ],
```
Comment (wangandi) [inline src/transform/src/predicate_pushdown.rs:682]: Nit: Another opportunity to use the MirScalarExpr::eq, et. al.

**Review by wangandi — APPROVED** [review]
🎉

File: src/transform/src/predicate_pushdown.rs (hunk 4 above)
Comment (ggevay) [reply src/transform/src/predicate_pushdown.rs]: Done

**Pull request merged**